File: test/clojure/bench/collection_churn.clj
;; Mirror of test/jacl/bench/collection_churn.jacl.
;;
;; Builds a 200-entry map and a 200-element vector, rebuilds each with
;; new values (functional/persistent updates), then sums all values
;; back via 200 reads each. Clojure's persistent collections are the
;; closest semantic match to JACL's HAMT/RRB, which makes this the
;; fairest cross-runtime comparison in the suite.
;; Validates: 2 * (sum 0..199 of i*2) = 2 * 39800 = 79600.

(ns bench.collection-churn)

(defn run []
  (let [n 200
        m1 (reduce (fn [m i] (assoc m i i)) {} (range n))
        m2 (reduce (fn [m i] (assoc m i (* i 2))) m1 (range n))
        ms (reduce + (map #(get m2 %) (range n)))
        v1 (reduce conj [] (range n))
        v2 (reduce (fn [v i] (assoc v i (* i 2))) v1 (range n))
        vs (reduce + (map #(nth v2 %) (range n)))]
    (+ ms vs)))

File: test/python/bench/collection_churn.py
"""Mirror of test/jacl/bench/collection_churn.jacl.

Builds a 200-entry dict and a 200-element list, rebuilds each with new
values, then sums all values back. Python has no persistent collections
in the stdlib, so this uses functional-style copy-on-write to match
JACL's semantics — every update produces a new collection. This makes
the comparison apples-to-apples on work done (alloc + copy) rather than
"how fast can mutable update go."

For an idiomatic mutable comparison, see collection_churn_mut.py.
"""


def run() -> int:
    n = 200

    m: dict[int, int] = {}
    for i in range(n):
        m = {**m, i: i}
    for i in range(n):
        m = {**m, i: i * 2}
    ms = 0
    for i in range(n):
        ms += m[i]

    v: list[int] = []
    for i in range(n):
        v = [*v, i]
    for i in range(n):
        v = [*v[:i], i * 2, *v[i + 1 :]]
    vs = 0
    for i in range(n):
        vs += v[i]

    return ms + vs


if __name__ == "__main__":
    print(run())
